framework/src/runtime.rs

```diff
@@ -0,0 +1,84 @@
+use allocators::CacheAligned;
+use common::Result;
+use config::{NetBricksConfiguration, CLI_ARGS};
+use interface::PortQueue;
+use scheduler::{initialize_system, NetBricksContext, StandaloneScheduler};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::prelude::{Future, Stream};
+use tokio::runtime::current_thread;
+use tokio::timer::Delay;
+use tokio_signal::unix::Signal;
+
+pub use tokio_signal::unix::{SIGHUP, SIGINT, SIGTERM};
+
+pub struct Runtime {
+    context: NetBricksContext,
+}
+
+impl Runtime {
+    /// Intializes the NetBricks context and starts the background schedulers
+    pub fn init(configuration: &NetBricksConfiguration) -> Result<Runtime> {
+        info!("initializing context:\n{}", configuration);
+        let mut context = initialize_system(configuration)?;
+        context.start_schedulers();
+        Ok(Runtime { context })
+    }
+
+    /// Runs a packet processing pipeline installer
+    pub fn add_pipeline_to_run<T>(&mut self, installer: T)
+    where
+        T: Fn(Vec<CacheAligned<PortQueue>>, &mut StandaloneScheduler) + Send + Sync + 'static,
+    {
+        self.context.add_pipeline_to_run(Arc::new(installer));
+    }
+
+    /// Executes tasks and pipelines
+    pub fn execute<T>(&mut self, on_signal: T) -> Result<()>
+    where
+        T: Fn(i32) -> std::result::Result<(), i32>,
+    {
+        let sighup = Signal::new(SIGHUP).flatten_stream();
+        let sigint = Signal::new(SIGINT).flatten_stream();
+        let sigterm = Signal::new(SIGTERM).flatten_stream();
+        let stream = sighup.select(sigint).select(sigterm);
+
+        let main_loop = stream.for_each(|signal| {
+            let code = match on_signal(signal) {
+                Ok(()) => 0,
+                Err(err) => err,
+            };
+
+            if signal != SIGHUP || code != 0 {
+                info!("shutting down context");
+                self.context.shutdown();
+                info!("exiting with code {}", code);
+                std::process::exit(code);
+            }
+
+            Ok(())
+        });
+
+        current_thread::block_on_all(main_loop).map_err(|e| e.into())
```

Comment:
@drunkirishcoder this makes sense to me too as nfs should be single-threaded and not need to spawn futures that impl Send. Are there any spots where we'd be wrong in forcing the single-threaded approach? This may be nice to discuss/document and put it up top on the module or here?

Reply:
I don't think this is settled. idea is runtime should be spun on the main process thread and is responsible for very light weight small tasks so a single thread shared for all the misc tasks should be fine. I want to look into if it's possible to pin this thread to a core and ensure it doesn't steal from other cores, especially those that do packet processing. if we end up using tokio more than for small misc tasks, then we will need to revisit this here. but for now this is ok. only other possible cpu intensive work I can think of that's in the plan right now is for example the connection cleanup timer, and I envision that will be its own dedicated thread possibly core as well separate from the main loop.

I will add more comments to the functions once the `Runtime` signature is settled.

Reply:
also, one nice benefit of running tokio in block mode on the main thread waiting for signals is we don't need to have an ugly sleep loop in `main`.

Reply:
sounds good.